react-ui/src/pages/entry/EntryModal.jsx

```diff
@@ -0,0 +1,349 @@
+  root: {
+    backgroundColor: theme.palette.common.black,
+  },
+  appBar: {
+    backgroundColor: 'inherit',
+  },
+  clickActive: {
+    cursor: 'pointer',
+  },
+  imageContainer: {
+    height: calculateImageContainerHeight(56),
+    position: 'relative',
+    [theme.breakpoints.up('sm')]: {
+      height: calculateImageContainerHeight(64),
+    },
+  },
+  image: {
+    maxHeight: '100%',
+    maxWidth: '100%',
+  },
+  navigateBefore: {
+    left: 28,
+  },
+  navigateButton: {
+    opacity: 0,
+    '&:focus': {
+      opacity: 1,
+    },
+  },
+  navigateNext: {
+    right: 28,
+  },
+  navigateVisible: {
+    opacity: 1,
+  },
+}));
+
+function EntryModal() {
+  /**
+   * Layout
+   */
+  const classes = useStyles();
+  const imageContainerRef = useRef(null);
+
+  /**
+   * AppBar
+   */
+  const redditCommentButtonRef = useRef(null);
+  const flagWaverButtonRef = useRef(null);
+
+  /**
+   * Info drawer
+   */
+  const [{ isInfoOpen }, updateSettings] = useSettingsState();
+  const isInfoOpenRef = useRef(isInfoOpen);
+  const updateInfoSetting = (infoOpen) => {
+    isInfoOpenRef.current = infoOpen;
+    updateSettings('isInfoOpen', infoOpen);
+  };
+  const handleDrawerClose = () => {
+    updateInfoSetting(false);
+  };
+
+  const toggleInfoDrawerOpen = () => {
+    updateInfoSetting(!isInfoOpenRef.current);
+  };
+
+  /**
+   * Entries
+   */
+  const { contestId, entryId } = useParams();
+  const apiPath = `/contests/${contestId}`;
+  const { data } = useSwrData(apiPath, false);
+  const {
+    entries = [],
+    winners = [],
+  } = data;
+  const allEntriesRef = useRef([]);
+  const [entry, setEntry] = useState({});
+  const [entryIndex, updateEntryIndex] = useState(-1);
+  const entryIndexRef = useRef(entryIndex);
+  const setEntryIndex = (value) => {
+    entryIndexRef.current = value;
+    updateEntryIndex(value);
+  };
+
+  /**
+   * Navigation
+   */
+  const navigate = useNavigate();
+  const { state = {} } = useLocation();
+  const [isNavigationAvailable, updateNavigationAvailable] = useState({
+    before: false,
+    next: false,
+  });
+  const isNavigationAvailableRef = useRef(isNavigationAvailable);
+  const setNavigationAvailable = (value) => {
+    isNavigationAvailableRef.current = value;
+    updateNavigationAvailable(value);
+  };
+  const [isNavigationVisible, setNavigationVisible] = useState({ before: true, next: true });
+  const hideNavigation = () => {
+    setNavigationVisible({ before: false, next: false });
+  };
+
+  const handleNavigate = (indexChange) => {
+    if (
+      !indexChange
+      || (indexChange < 0 && !isNavigationAvailableRef.current.before)
+      || (indexChange > 0 && !isNavigationAvailableRef.current.next)
+    ) {
+      hideNavigation();
+      return;
+    }
+
+    setEntry({});
+
+    const { id } = allEntriesRef.current[entryIndexRef.current + indexChange];
+    navigate(`../${id}`, { relative: 'path', replace: true, state });
+  };
+
+  const handleKeyUp = ({ key }) => {
```

Comment:
Nit: This isn't strictly for navigation, perhaps it should move to a different section or later after the navigation-related code?

Reply:
It does rely on the code immediately above and is used immediately below, but I forget how React handles function order

Reply:
Ah, I think it wants the function to be defined before it's called, so makes sense to keep it here then.